### PR TITLE
Use caching logic to determine openapi spec

### PR DESCRIPTION
**Pull request by dmontagu**
Dependencies that resolve to the same cache key are not re-used while building the openapi schema.

Addresses #416

### CODE REVIEW

**Comment by dmontagu** [comment]
There's a nuance here related to security -- as implemented, if you were to have the same callable in multiple places with different security, you'd get repeats of the dependency in the schema.

In theory, this could be avoided if there was a way to "merge" all of the relevant security scopes. I'm not sure whether that is possible, and it seems like enough of an edge case to not handle for now.

File: tests/test_repeated_dependency_schema.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
+from fastapi import Depends, FastAPI, Header
+from starlette.status import HTTP_200_OK
+from starlette.testclient import TestClient
+
+app = FastAPI(title="Access", redoc_url=None)
```
Comment (henriklindgren) [inline tests/test_repeated_dependency_schema.py]: could remove title and redoc arguments here that slipped in from my project, sorry :)

File: tests/test_repeated_dependency_schema.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+                "summary": "Get Deps",
+            }
+        }
+    },
+}
+
+
+def test_schema():
+    response = client.get("/openapi.json")
+    assert response.status_code == 200
```
Comment (henriklindgren) [inline tests/test_repeated_dependency_schema.py]: could use HTTP_200_OK here too

**Comment by henriklindgren** [comment]
@dmontagu have you considered if this affects https://fastapi.tiangolo.com/tutorial/header-params/#duplicate-headers?

**Comment by dmontagu** [comment]
@henriklindgren That's a good question, but I think this is still safe -- it only modifies things when calling the exact same dependency function. If you were to list a header argument multiple times in *different* dependencies, it would still get added multiple times.

For what it's worth, if I understand correctly, any schema that would be changed by this is probably invalid already anyway.

**Comment by henriklindgren** [comment]
Ah true, read through it all now, agreed. The way I noticed the bug was from generating a client which broke from it, I would be surprised if people would use it this way on purpose so it should be safe.

>  if you were to have the same callable as a dependency in multiple places with different security schemes, you'd get a repeat of the dependency's arguments in the schema for each different security scheme

Let's see if I've understood this part correctly. It seems safe to ignore this case because it would be really awkward if multiple security schemes in the same context based key values on the order they arrived, dependent on other schemes. It honestly feels safer to know that you would get the single argument and be sure of it. Otherwise you'd use argument lists as mentioned (such as the duplicate header list), which could be used as a workaround if you really want this behaviour.

**Comment by tiangolo** [comment]
Great! Thanks @dmontagu for the thorough investigation and explanation, here and in the issue. :cake: :rocket:

---

I think we can have this merged as is right now.

Combining the scopes of the same security schemes could make sense, but that would mean "in OpenAPI terms" that *all* the scopes are required. And the intention could be to have *any* of them required. That's actually not enforced by FastAPI but by the implementation itself. If the user/developer leaves the `auto_error=True`, then *all* the security schemes would be required. But if he has `auto_error=False` and then some additional logic to decide if the operation is allowed, then it could be a different case.

This is a pretty advanced use case of OpenAPI, having multiple security schemes combined that some can be required or not.

I think we can leave it as-is (with this PR merged) for now and see if there's more logic required for that for some use case later.
